webapp/src/ts/services/training-cards.service.ts

```diff
@@ -0,0 +1,124 @@
+import * as moment from 'moment';
+
+import { XmlFormsService } from '@mm-services/xml-forms.service';
+import { TrainingCardsComponent } from '@mm-modals/training-cards/training-cards.component';
+import { DbService } from '@mm-services/db.service';
+import { GlobalActions } from '@mm-actions/global';
+import { ModalService } from '@mm-modals/mm-modal/mm-modal';
+import { SessionService } from '@mm-services/session.service';
+import { RouteSnapshotService } from '@mm-services/route-snapshot.service';
+
+export const TRAINING_PREFIX: string = 'training';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class TrainingCardsService {
+  private globalActions;
+
+  constructor(
+    private store: Store,
+    private xmlFormsService: XmlFormsService,
+    private dbService: DbService,
+    private modalService: ModalService,
+    private sessionService: SessionService,
+    private routeSnapshotService: RouteSnapshotService,
+  ) {
+    this.globalActions = new GlobalActions(store);
+  }
+
+  private getAvailableTrainingForms(xForms, userCtx) {
+    const today = moment();
+
+    return xForms
+      .map(xForm => ({
+        id: xForm._id,
+        code: xForm.internalId,
+        startDate: xForm.context?.start_date,
+        duration: xForm.context?.duration,
+        userRoles: xForm.context?.user_roles,
+      }))
+      .filter(form => {
+        const hasRole = form.userRoles?.find(role => userCtx.roles.includes(role));
+        if (!hasRole) {
+          return false;
+        }
+
+        form.startDate = form.startDate ? moment(form.startDate) : today.clone();
+        if (!form.startDate?.isValid() || form.startDate.isAfter(today)) {
+          return false;
+        }
+
+        if (!form.duration) {
+          return true; // Training never ends
+        }
+
+        const endDate = form.startDate.clone();
+        endDate.add(form.duration, 'day');
+        return endDate.isAfter(today);
+      });
+  }
+
+  private async getCompletedTrainings(userCtx) {
+    const docs = await this.dbService
+      .get()
+      .allDocs({
+        include_docs: true,
+        startkey: [ TRAINING_PREFIX, userCtx.name ].join(':'),
+        endkey: [ TRAINING_PREFIX, userCtx.name, '\ufff0' ].join(':'),
+      });
+
+    return docs?.rows?.length ? new Set(docs.rows.map(row => row?.doc?.form)) : new Set();
+  }
+
+  private async handleTrainingCards(error, xForms) {
+    if (error) {
+      console.error('Error fetching training cards.', error);
+      return;
+    }
+
+    const routeSnapshot = this.routeSnapshotService.get();
+    if (routeSnapshot?.data?.hideTraining) {
+      return;
+    }
+
+    try {
+      const firstChronologicalTrainingForm = await this.getFirstChronologicalForm(xForms);
+      if (!firstChronologicalTrainingForm) {
+        return;
+      }
+
+      this.globalActions.setTrainingCard(firstChronologicalTrainingForm.code);
+      this.modalService.show(TrainingCardsComponent, { backdrop: 'static' });
```

Comment:
I get an uncaught error logged in the console if I press escape when the modal window is open. 
```javascript
feedback.service.ts:79 Error: Uncaught (in promise): [object Undefined]
    at resolvePromise (zone.js:1211:31)
    at zone.js:1118:17
    at Object.reject (zone.js:1134:33)
    at TrainingCardsComponent.modalReject (mm-modal.ts:108:62)
    at TrainingCardsComponent.cancel (mm-modal.ts:98:10)
    at Object.next (mm-modal.ts:65:47)
    at ConsumerObserver.next (Subscriber.js:91:33)
    at SafeSubscriber._next (Subscriber.js:60:26)
    at SafeSubscriber.next (Subscriber.js:31:18)
    at take.js:12:32
```
This also creates a feedback document. 
This also happens if I go back in browser history when the modal is open.

Reply:
It shouldn't close if you press the Escape keyboard. It's a regression, fixing it now

Reply:
Going back in browser history (another way of forcefully closing) shouldn't log an error or create a feedback doc either.

Reply:
Fixed